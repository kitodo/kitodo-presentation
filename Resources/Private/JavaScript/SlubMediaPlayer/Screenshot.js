// @ts-check

/**
 * @typedef {{
 *  h: 'left' | 'right';
 *  v: 'top' | 'bottom';
 *  text: string;
 * }} ScreenshotCaption
 *
 * @typedef {{
 *  captions: ScreenshotCaption[];
 *  minWidth: number;
 * }} ScreenshotConfig
 */

/**
 *
 * @param {HTMLCanvasElement | CanvasRenderingContext2D} target Canvas to which
 * the screenshot is drawn
 * @param {HTMLVideoElement} videoDomElement Source video element from which
 * the screenshot is taken
 * @param {Partial<ScreenshotConfig>} config
 * @returns {boolean}
 */
export function drawScreenshot(target, videoDomElement, config) {
  const [targetCanvas, context] =
    target instanceof HTMLCanvasElement
      ? [target, target.getContext('2d')]
      : [target.canvas, target];

  if (context === null) {
    return false;
  }

  // Make sure the target resolution is a multiple of the video resolution
  const targetFactor =
    Math.max(1, Math.ceil((config.minWidth ?? 0) / videoDomElement.videoWidth));

  targetCanvas.width = videoDomElement.videoWidth * targetFactor;
  targetCanvas.height = videoDomElement.videoHeight * targetFactor;

  context.drawImage(
    videoDomElement,
    0, 0, targetCanvas.width, targetCanvas.height
  );

  const unitHeight = targetCanvas.height / 1080;
  const textPad = 10 * unitHeight;

  context.font = `${Math.floor(25 * unitHeight)}px Arial`;
  context.fillStyle = "#FFFFFF";
  context.shadowBlur = 5;
  context.shadowColor = "black";

  for (const caption of config.captions ?? []) {
    const x = caption.h === 'left' ? textPad : targetCanvas.width - textPad;
    const y = caption.v === 'top' ? textPad : targetCanvas.height - textPad;

    context.textAlign = caption.h;
    context.fillText(caption.text, x, y);
  }

  return true;
}
