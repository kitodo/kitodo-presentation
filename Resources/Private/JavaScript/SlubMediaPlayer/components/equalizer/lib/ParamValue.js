// @ts-check

/**
 * @implements {dlf.media.eq.ParamValue}
 */
export default class ParamValue {
  /**
   *
   * @param {number} value
   * @param {((oldValue: number, newValue: number, override: (value: number) => void) => void) | null} updateFn
   */
  constructor(value, updateFn = null) {
    /** @private */
    this.initial_ = value;

    /** @private */
    this.value_ = value;

    /** @private */
    this.updateFn_ = updateFn;
  }

  get initial() {
    return this.initial_;
  }

  get value() {
    return this.value_;
  }

  get editable() {
    return this.updateFn_ !== null;
  }

  /**
   * TODO?
   *
   * @param {number} value
   */
  initialize(value) {
    this.initial_ = value;
    this.value_ = value;
  }

  /**
   *
   * @param {(oldValue: number) => number} fn
   */
  update(fn) {
    if (this.updateFn_ !== null) {
      const oldValue = this.value_;
      const newValue = fn(oldValue);
      this.value_ = newValue;
      this.updateFn_(oldValue, newValue, (value) => {
        this.value_ = value;
      });
    }
  }
}
