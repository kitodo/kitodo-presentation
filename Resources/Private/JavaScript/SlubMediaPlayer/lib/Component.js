// @ts-check

import EventEmitter from 'events';

/**
 * @template {object} State
 */
export default class Component extends EventEmitter {
  /**
   *
   * @param {State} state
   */
  constructor(state) {
    super();

    /**
     * @protected
     * @type {State}
     */
    this.state = state;

    /**
     * @private
     * @type {((prevState: State) => Partial<State>)[]}
     */
    this.pendingStateUpdates = [];

    /**
     * @private
     * @type {ReturnType<setTimeout> | null}
     */
    this.renderTimeout = null;

    /**
     * @private
     * @type {Promise<void>}
     */
    this.renderPromise = Promise.resolve();
  }

  /**
   *
   * @param {Partial<State> | ((prevState: State) => Partial<State>)} state
   */
  setState(state = {}) {
    const stateFn = typeof state === 'function' ? state : (() => state);
    this.pendingStateUpdates.push(stateFn);

    // Postpone updates so that multiple synchronous calls to `setState` don't
    // lead to multiple renderings.
    if (!this.renderTimeout) {
      this.renderPromise = new Promise((resolve) => {
        this.renderTimeout = setTimeout(() => {
          const newState = this.squashStateUpdates();
          this.render(newState);
          this.state = newState;
          this.renderTimeout = null;
          this.renderPromise = Promise.resolve();
          this.emit('updated', newState);
          resolve();
        });
      });
    }
  }

  /**
   * Returns a promise of any pending rerender being completed. (If no rerender
   * is pending, the returned promise is already resolved.)
   *
   * @returns {Promise<void>}
   */
  update() {
    return this.renderPromise;
  }

  /**
   * @private
   * @returns {State}
   */
  squashStateUpdates() {
    const newState = Object.assign({}, this.state);
    for (const updateState of this.pendingStateUpdates) {
      Object.assign(newState, updateState(newState));
    }
    this.pendingStateUpdates = [];
    return newState;
  }

  /**
   * Rerenders the component based on new state.
   *
   * `this.state` still refers to the old state during execution of this method,
   * which you may use to detect state changes.
   *
   * @abstract
   * @protected
   * @param {State} state The updated state
   */
  render(state) {
    //
  }
}
