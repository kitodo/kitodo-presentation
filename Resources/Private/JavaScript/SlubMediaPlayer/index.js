// @ts-check

import 'abortcontroller-polyfill/dist/polyfill-patch-fetch';

import SlubMediaPlayer from 'SlubMediaPlayer/SlubMediaPlayer';
window.SlubMediaPlayer = SlubMediaPlayer;

import 'SlubMediaPlayer/components/marker-table';
import 'SlubMediaPlayer/components/equalizer/EqualizerPlugin';
