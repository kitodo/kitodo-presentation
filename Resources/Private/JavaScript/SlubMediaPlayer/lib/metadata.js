// @ts-check

import { getTimeStringPlaceholders } from 'DlfMediaPlayer/index';
import { fillPlaceholders } from 'lib/util';

/**
 * In {@link template} string, fill placeholders representing metadata with the
 * value given in {@link metadata}.
 *
 * @param {string} template
 * @param {MetadataArray} metadata
 * @returns {string}
 */
export function fillMetadata(template, metadata) {
  const firstMetadataValues = Object.fromEntries(
    Object.entries(metadata).map(([key, values]) => [key, values[0] ?? ''])
  );

  return fillPlaceholders(template, firstMetadataValues);
}

/**
 * Create metadata array that extends {@link metadata} and contains additional
 * information that may be relevant for filling placeholders.
 *
 * The result may be used, in particular, with {@link fillMetadata}.
 *
 * @param {dlf.media.UrlGenerator} gen
 * @param {MetadataArray} metadata
 * @param {number | dlf.media.TimeRange | null} timecode
 * @param {number | null} fps
 * @returns {MetadataArray}
 */
export function makeExtendedMetadata(gen, metadata, timecode, fps) {
  const url = gen.generateUrl(timecode);

  /** @type {MetadataArray} */
  const result = {
    ...metadata,
    host: [`${location.protocol}//${location.host}`],
    url: [url.toString()],
  };

  // TODO: What's actually a good behavior when timecode is missing?
  let startTime = 0;
  if (typeof timecode === 'number') {
    startTime = timecode;
  } else if (timecode !== null) {
    startTime = timecode.startTime;
  }
  const timeStringPlaceholders = getTimeStringPlaceholders(startTime, fps);
  for (const [key, value] of Object.entries(timeStringPlaceholders)) {
    result[key] = [value];
  }

  return result;
}
