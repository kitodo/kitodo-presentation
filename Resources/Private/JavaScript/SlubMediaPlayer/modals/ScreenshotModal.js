// @ts-check

import {
  binaryStringToArrayBuffer,
  blobToBinaryString,
  canvasToBlob,
  download,
  e,
  domJoin,
  sanitizeBasename,
} from 'lib/util';
import imageFormats from 'lib/image/imageFormats';
import typoConstants from 'lib/typoConstants';
import { fillMetadata, makeExtendedMetadata } from 'SlubMediaPlayer/lib/metadata';
import { drawScreenshot } from 'SlubMediaPlayer/Screenshot';
import { getKeybindingText } from 'SlubMediaPlayer/lib/trans';
import SimpleModal from 'SlubMediaPlayer/lib/SimpleModal';
import UrlGenerator from 'SlubMediaPlayer/lib/UrlGenerator';

/**
 * @typedef {{
 *  metadata: MetadataArray | null;
 *  showMetadata: boolean;
 *  fps: number | null;
 *  timecode: number | null;
 *  supportedImageFormats: ImageFormatDesc[];
 *  selectedImageFormat: ImageFormatDesc | null;
 * }} State
 *
 * @typedef {{
 *  keybindings: Keybinding<any, any>[];
 *  screnshotCaptions: import('SlubMediaPlayer/Screenshot').ScreenshotCaption[];
 *  constants: import('lib/typoConstants').TypoConstants<ScreenshotModalConstants>;
 * }} Config
 */

/**
 * @extends {SimpleModal<State>}
 */
export default class ScreenshotModal extends SimpleModal {
  /**
   *
   * @param {HTMLElement} parent
   * @param {Translator & Identifier & Browser} env
   * @param {Config} config
   */
  constructor(parent, env, config) {
    const supportedImageFormats = imageFormats.filter(
      format => env.supportsCanvasExport(format.mimeType)
    );

    super(parent, {
      metadata: null,
      showMetadata: true,
      fps: null,
      timecode: null,
      supportedImageFormats,
      selectedImageFormat: supportedImageFormats[0] ?? null,
    });

    /** @private */
    this.env = env;
    /** @private */
    this.gen = new UrlGenerator(env);
    /** @private @type {HTMLVideoElement | null} */
    this.videoDomElement = null;
    /** @private */
    this.config = config;
    /** @private */
    this.constants = typoConstants(config.constants, {
      screenshotFilenameTemplate: 'Screenshot',
      screenshotCommentTemplate: '',
    });

    const snapKeybinding = this.config.keybindings.find(
      kb => kb.action === 'modal.screenshot.snap'
    );

    this.$main.classList.add('screenshot-modal');
    this.$title.innerText = env.t('modal.screenshot.title');

    const idShowMetadata = env.mkid();
    const radioGroup = env.mkid();

    this.$body.append(
      e("div", { className: "screenshot-config" }, [
        e("h4", {}, [env.t('modal.screenshot.configuration')]),
        e("section", { className: "metadata-config" }, [
          e("h1", {}, [env.t('modal.screenshot.metadata')]),
          e("div", { className: "metadata-overlay" }, [
            e("input", {
              type: "checkbox",
              id: idShowMetadata,
              checked: this.state.showMetadata,
              $change: this.handleChangeShowMetadata.bind(this),
            }),
            e("label", { htmlFor: idShowMetadata }, [
              env.t('modal.screenshot.metadata-overlay'),
            ]),
          ]),
        ]),
        e("section", {}, [
          e("h1", {}, [env.t('modal.screenshot.file-format')]),
          e("div", {}, (
            this.state.supportedImageFormats.map(format => {
              const radioId = env.mkid();

              return e("span", { className: "file-format-option" }, [
                e("input", {
                  id: radioId,
                  name: radioGroup,
                  type: 'radio',
                  checked: format.mimeType === this.state.selectedImageFormat?.mimeType,
                  $change: () => {
                    this.setState({
                      selectedImageFormat: format,
                    });
                  },
                }),
                e("label", { htmlFor: radioId }, [` ${format.label}`]),
              ]);
            })
          )),
        ]),
        e("a", {
          href: "#",
          className: "download-link",
          $click: this.handleDownloadImage.bind(this),
        }, [
          e("i", {
            className: "material-icons-round inline-icon",
          }, ["download"]),
          env.t('modal.screenshot.download-image'),
        ]),
        snapKeybinding && (
          e("aside", { className: "snap-tip" }, [
            e("i", {
              className: "material-icons-round inline-icon",
            }, ["info_outline"]),
            e("span", {}, (
              domJoin(
                env.t('modal.screenshot.snap-tip', { keybinding: "{kb}" }).split('{kb}'),
                getKeybindingText(env, snapKeybinding)
              )
            )),
          ])
        ),
      ]),

      this.$canvas = e("canvas")
    );
  }

  /**
   * Sets video DOM element for upcoming screenshots.
   *
   * @param {HTMLVideoElement} video
   * @returns {this}
   */
  setVideo(video) {
    this.videoDomElement = video;
    return this;
  }

  /**
   * Triggers UI update using new {@link metadata}.
   *
   * @param {MetadataArray} metadata
   * @returns {this}
   */
  setMetadata(metadata) {
    this.setState({ metadata });
    return this;
  }

  /**
   * Triggers UI update using new {@link fps}.
   *
   * @param {number | null} fps
   * @returns {this}
   */
  setFps(fps) {
    this.setState({ fps });
    return this;
  }

  /**
   * Triggers UI update using new {@link timecode}.
   *
   * @param {number} timecode
   * @returns {this}
   */
  setTimecode(timecode) {
    this.setState({ timecode });
    return this;
  }

  /**
   * @private
   * @param {Event} e
   */
  handleChangeShowMetadata(e) {
    if (!(e.target instanceof HTMLInputElement)) {
      return;
    }

    this.setState({
      showMetadata: e.target.checked,
    });
  }

  /**
   * @private
   * @param {MouseEvent} e
   */
  async handleDownloadImage(e) {
    e.preventDefault();

    // We could've set `.download-image[href]` in `render()` or in the radio
    // box change listener, but avoid this for performance reasons.

    await this.downloadCurrentImage(this.state);
  }

  /**
   * @param {Pick<State, 'showMetadata' | 'metadata' | 'timecode' | 'fps'>} state
   */
  renderCurrentScreenshot({ showMetadata, metadata, timecode, fps }) {
    if (this.videoDomElement === null) {
      // TODO: Error handling
      return false;
    }

    const config = {
      /** @type {import('SlubMediaPlayer/Screenshot').ScreenshotCaption[]} */
      captions: [],
      minWidth: 1000,
    };

    if (showMetadata) {
      const extendedMetadata = makeExtendedMetadata(this.gen, metadata ?? {}, timecode, fps);
      config.captions = this.getCaptions(extendedMetadata);
    }

    drawScreenshot(this.$canvas, this.videoDomElement, config);

    return true;
  }

  /**
   *
   * @param {Pick<State, 'metadata'| 'fps' | 'timecode' | 'selectedImageFormat'>} state
   */
  async downloadCurrentImage(state) {
    const { metadata, timecode, fps, selectedImageFormat } = state;
    if (metadata === null || timecode === null || selectedImageFormat === null) {
      console.error("one of [metadata, timecode, selectedImageFormat] is null");
      return false;
    }

    const extendedMetadata = makeExtendedMetadata(this.gen, metadata, timecode, fps);

    const image = await this.makeImageBlob(this.$canvas, selectedImageFormat, extendedMetadata);
    const filename = this.getFilename(extendedMetadata, selectedImageFormat);

    download(image, filename);

    return true;
  }

  /**
   *
   * @param {HTMLCanvasElement} canvas
   * @param {ImageFormatDesc} imageFormat
   * @param {MetadataArray} metadata
   */
  async makeImageBlob(canvas, imageFormat, metadata) {
    const imageBlob = await canvasToBlob(canvas, imageFormat.mimeType);
    const imageDataStr = await blobToBinaryString(imageBlob);
    const image = imageFormat.parseBinaryString(imageDataStr);

    if (image) {
      image.addMetadata({
        title: metadata.title?.[0] ?? "",
        // NOTE: Don't localize (not only relevant to current user)
        comment: fillMetadata(this.constants.screenshotCommentTemplate, metadata),
      });
      const buffer = binaryStringToArrayBuffer(image.toBinaryString());
      return new Blob([buffer], { type: imageBlob.type });
    } else {
      return imageBlob;
    }
  }

  /**
   *
   * @param {MetadataArray} metadata
   * @param {ImageFormatDesc} selectedImageFormat
   * @return {string}
   */
  getFilename(metadata, selectedImageFormat) {
    const template = this.constants.screenshotFilenameTemplate;
    const basename = fillMetadata(template, metadata);
    const extension = selectedImageFormat.extension;

    return `${sanitizeBasename(basename)}.${extension}`;
  }

  /**
   *
   * @param {MetadataArray} metadata
   * @returns {import('SlubMediaPlayer/Screenshot').ScreenshotCaption[]}
   */
  getCaptions(metadata) {
    return this.config.screnshotCaptions.map(caption => ({
      ...caption,
      text: fillMetadata(caption.text, metadata),
    }));
  }

  /**
   * Downloads image without opening the modal.
   */
  async snap() {
    // Parameters may be on the way via setState (TODO: Refactor)
    await this.update();

    const state = this.state;
    const success = (
      this.renderCurrentScreenshot(state)
      && await this.downloadCurrentImage(state)
    );

    if (!success) {
      alert(this.env.t('modal.screenshot.error'));
    }
  }

  /**
   * @override
   * @param {import('SlubMediaPlayer/lib/SimpleModal').BaseState & State} state
   */
  render(state) {
    super.render(state);

    const shouldRender = (
      state.show
      && (!this.state.show || state.showMetadata !== this.state.showMetadata)
    );

    if (shouldRender) {
      this.renderCurrentScreenshot(state);
    }
  }
}
