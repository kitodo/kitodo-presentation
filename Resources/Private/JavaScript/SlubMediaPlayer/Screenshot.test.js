/**
 * @jest-environment jsdom
 */

// @ts-check

import { beforeEach, expect, test } from '@jest/globals';
import Environment from 'lib/Environment';
import { fillMetadata } from 'SlubMediaPlayer/lib/metadata';
import { drawScreenshot } from 'SlubMediaPlayer/Screenshot';
import ScreenshotModal from 'SlubMediaPlayer/modals/ScreenshotModal';

class VideoMock extends HTMLVideoElement {
  /**
   * @param {number} width
   * @param {number} height
   */
  constructor(width, height) {
    super();

    this._mockWidth = width;
    this._mockHeight = height;
  }

  /** @override */
  get width() {
    return this._mockWidth;
  }

  /** @override */
  get videoWidth() {
    return this._mockWidth;
  }

  /** @override */
  get height() {
    return this._mockHeight;
  }

  /** @override */
  get videoHeight() {
    return this._mockHeight;
  }
}

customElements.define('video-mock', VideoMock, { extends: 'video' });

beforeEach(() => {
  // TODO: Reset JSDOM in a more robust way
  document.body.innerHTML = '';
});

function getEnvironment() {
  const env = new Environment();
  env.setLang({
    locale: 'en_US',
    twoLetterIsoCode: 'en',
    phrases: {
      'key.generic': "{key}",
      'key.unto': " to ",
      'modal.screenshot.title': "Screenshot",
      'modal.screenshot.configuration': "Show Metadata",
      'modal.screenshot.download-image': "Download Image",
      'modal.screenshot.metadata': "Metadata",
      'modal.screenshot.metadata-overlay': "Overlay on image",
      'modal.screenshot.file-format': "File Format",
      'modal.screenshot.snap-tip': "Tip: Snap via {keybinding}",
    },
  });
  return env;
}

function getTestMetadataArray() {
  return {
    title: ["Test Video"],
    year: ["1912"],
  };
}

test('can open screenshot overlay', async () => {
  const overlay = () => document.querySelector('.screenshot-modal');

  // not opened yet
  expect(overlay()).toBeNull();

  // opened; exact tags are in snapshot
  const modal = new ScreenshotModal(document.body, getEnvironment(), {
    keybindings: [
      {
        keys: ["s"],
        action: "modal.screenshot.snap",
        kind: "other",
        order: 0
      },
    ],
    screnshotCaptions: [],
    constants: {},
  });
  modal.setVideo(new VideoMock(1920, 1080));
  modal.setMetadata(getTestMetadataArray());
  modal.open();
  await modal.update();
  expect(overlay()).toMatchSnapshot();
});

test('can draw to canvas', () => {
  const metadata = getTestMetadataArray();

  /**
   * @param {number} videoWidth
   * @param {number} videoHeight
   * @param {number} minWidth
   */
  const snapshotWithSize = (videoWidth, videoHeight, minWidth = 0) => {
    const video = new VideoMock(videoWidth, videoHeight);

    const canvas = document.createElement("canvas");
    const context = canvas.getContext('2d');

    if (context === null) {
      throw new Error();
    }

    drawScreenshot(context, video, {
      captions: [
        { v: 'top', h: 'left', text: "top left" },
        { v: 'top', h: 'right', text: "top right" },
        { v: 'bottom', h: 'left', text: "bottom left" },
        { v: 'bottom', h: 'right', text: fillMetadata("{title} / {year}", metadata) },
      ],
      minWidth,
    });

    // @ts-ignore TODO: Why wouldn't it recognize "__getEvents"?
    const events = context.__getEvents();
    expect(events).toMatchSnapshot();

    expect(canvas.width).toBeGreaterThanOrEqual(minWidth);
    expect(canvas.width / canvas.height).toBeCloseTo(videoWidth / videoHeight);
  };

  snapshotWithSize(1920, 1080);
  snapshotWithSize(960, 540);
  snapshotWithSize(320, 180, /* minWidth= */ 480);
});
