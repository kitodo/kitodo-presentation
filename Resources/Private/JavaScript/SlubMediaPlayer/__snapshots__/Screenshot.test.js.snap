// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can draw to canvas 1`] = `
Array [
  Object {
    "props": Object {
      "dHeight": 1080,
      "dWidth": 1920,
      "dx": 0,
      "dy": 0,
      "img": VideoMock {
        "_mockHeight": 1080,
        "_mockWidth": 1920,
      },
      "sHeight": 1080,
      "sWidth": 1920,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "value": "25px Arial",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "#ffffff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "value": 5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "shadowBlur",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "shadowColor",
  },
  Object {
    "props": Object {
      "value": "left",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "top left",
      "x": 10,
      "y": 10,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "value": "right",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "top right",
      "x": 1910,
      "y": 10,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "value": "left",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "bottom left",
      "x": 10,
      "y": 1070,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "value": "right",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "Test Video / 1912",
      "x": 1910,
      "y": 1070,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
]
`;

exports[`can draw to canvas 2`] = `
Array [
  Object {
    "props": Object {
      "dHeight": 540,
      "dWidth": 960,
      "dx": 0,
      "dy": 0,
      "img": VideoMock {
        "_mockHeight": 540,
        "_mockWidth": 960,
      },
      "sHeight": 540,
      "sWidth": 960,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "value": "12px Arial",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "#ffffff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "value": 5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "shadowBlur",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "shadowColor",
  },
  Object {
    "props": Object {
      "value": "left",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "top left",
      "x": 5,
      "y": 5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "value": "right",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "top right",
      "x": 955,
      "y": 5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "value": "left",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "bottom left",
      "x": 5,
      "y": 535,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "value": "right",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "Test Video / 1912",
      "x": 955,
      "y": 535,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
]
`;

exports[`can draw to canvas 3`] = `
Array [
  Object {
    "props": Object {
      "dHeight": 180,
      "dWidth": 320,
      "dx": 0,
      "dy": 0,
      "img": VideoMock {
        "_mockHeight": 180,
        "_mockWidth": 320,
      },
      "sHeight": 180,
      "sWidth": 320,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "value": "8px Arial",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "#ffffff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "value": 5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "shadowBlur",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "shadowColor",
  },
  Object {
    "props": Object {
      "value": "left",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "top left",
      "x": 3.333333333333333,
      "y": 3.333333333333333,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "value": "right",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "top right",
      "x": 636.6666666666666,
      "y": 3.333333333333333,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "value": "left",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "bottom left",
      "x": 3.333333333333333,
      "y": 356.6666666666667,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "value": "right",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "Test Video / 1912",
      "x": 636.6666666666666,
      "y": 356.6666666666667,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
]
`;

exports[`can open screenshot overlay 1`] = `
<div
  class="sxnd-modal screenshot-modal"
>
  <div
    class="headline-container"
  >
    <h3 />
    <span
      class="modal-close material-icons-round"
    >
      close
    </span>
  </div>
  <div
    class="body-container"
    style="max-height: calc(0px - 11rem);"
  >
    <div
      class="screenshot-config"
    >
      <h4>
        Show Metadata
      </h4>
      <section
        class="metadata-config"
      >
        <h1>
          Metadata
        </h1>
        <div
          class="metadata-overlay"
        >
          <input
            id="__autoid_1"
            type="checkbox"
          />
          <label
            for="__autoid_1"
          >
            Overlay on image
          </label>
        </div>
      </section>
      <section>
        <h1>
          File Format
        </h1>
        <div>
          <span
            class="file-format-option"
          >
            <input
              id="__autoid_3"
              name="__autoid_2"
              type="radio"
            />
            <label
              for="__autoid_3"
            >
               PNG
            </label>
          </span>
          <span
            class="file-format-option"
          >
            <input
              id="__autoid_4"
              name="__autoid_2"
              type="radio"
            />
            <label
              for="__autoid_4"
            >
               JPEG
            </label>
          </span>
        </div>
      </section>
      <a
        class="download-link"
        href="#"
      >
        <i
          class="material-icons-round inline-icon"
        >
          download
        </i>
        Download Image
      </a>
      <aside
        class="snap-tip"
      >
        <i
          class="material-icons-round inline-icon"
        >
          info_outline
        </i>
        <span>
          Tip: Snap via 
          <span>
            <kbd>
              S
            </kbd>
          </span>
          
        </span>
      </aside>
    </div>
    <canvas
      height="1080"
      width="1920"
    />
  </div>
</div>
`;
